components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -93,6 +94,49 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes and reads here.
+    protected readonly contentCache = new Map<string, string>();
+    protected updatedSinceLastWrite = new Set<string>();
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.add(resource.uri);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+
+        const content = await this.getServer().getUserStorageResource(req);
+        this.contentCache.set(req.uri, content);
+        return content;
+    }
+
+    protected startPeriodicCacheWriter(interval: number) {
+        const performWrites = async () => {
+            const writeStart = Date.now();
+
+            const writtenUris = this.updatedSinceLastWrite;
+            this.updatedSinceLastWrite = new Set<string>();
+
+            try {
+                const server = this.getServer();
+                for (const uri of writtenUris) {
+                    const content = this.contentCache.get(uri)!;
+                    await server.updateUserStorageResource({ uri, content });
+                }
+            } catch (err) {
+                console.error(err);
```

Comment:
if there was a reconnection then we should resubmit URIs. You can check for reconnections with `server.onDidCloseConnection`.

Reply:
https://github.com/gitpod-io/gitpod/pull/2476/commits/41caa1b321d014846f96a7c48c101b3c6d2b9ae7